### PR TITLE
Inverse evaluation update

### PR DESCRIPTION
**Pull request by qnzhou**
### Summary:
* Bug fix in 2nd derivative evaluation of `duv` for periodic patch.
* Minor change in subpatch-based inverse evaluation to trim the domain first.  This trimming is important because it can cut off out-of-domain singularities, which were causing so much headache before.
* Minor performance tuning.  Naive bisection algorithm now uses different sampling rate in u and v directions.
* Update inverse evaluation API (_warning: breaking change_), see below for more detail.
* As always, add more unit tests.

Here is a quick summary of the new patch inverse evaluation API:

#### Version 1:
```c++
std::tie(uv, converged) = patch.inverse_evaluate(
    query_point,               ///< Query point
    u_min, u_max, v_min, v_max ///< Search domain
);
```
The arguments are the same as before, the query point and the search domain.  The return value is now a tuple containing the `uv` coordinates of the closest point and a `converged` flag indicating whether the Newton iterations have converged.

Note that a converged solution does _not_ mean we have successfully found the closest point.  It is possible the result is just a local minimum.

#### Version 2:
```c++
std::tie(uv, converged) = patch.inverse_evaluate(
    query_point,                  ///< Query point
    initial_u, initial_v,         ///< Initial guess, ideally sufficiently close to the solution.
    u_min, u_max, v_min, v_max,   ///< Search domain
    num_iterations,               ///< Optional, number of Newton iterations to run.
    tolerence                     ///< Options, convergence tolerence.
);
```
In this version, we take an initial guess of the solution and run Newton iterations starting from it.  A few optional parameters to the Newton iteration is also supported in the API.  The result values are the same as before.  This version is potentially faster since it skips the expensive bisection or subpatch algorithms used to find a good initial guess in the previous version.

### CODE REVIEW

File: tests/test_BSplinePatch.cpp [hunk 1]
```diff
@@ -266,15 +266,15 @@ TEST_CASE("BSplinePatch", "[nonrational][bspline_patch]")
...
-        const auto uv1 = patch.inverse_evaluate(q, u_max - d, u_max + d, v_min, v_max);
+        const auto uv1 = std::get<0>(patch.inverse_evaluate(q, u_max - d, u_max + d, v_min, v_max));
         REQUIRE((patch.evaluate(uv1[0], uv1[1]) - q).norm() == Approx(0));
         const auto uv2 =
-            patch.inverse_evaluate(q, u_min - 2 * period - d, u_min - 2 * period + d, v_min, v_max);
+            std::get<0>(patch.inverse_evaluate(q, u_min - 2 * period - d, u_min - 2 * period + d, v_min, v_max));
         REQUIRE((patch.evaluate(uv2[0], uv2[1]) - q).norm() == Approx(0));
         const auto uv3 =
-            patch.inverse_evaluate(q, u_min - period + d, u_max - period - d, v_min, v_max);
+            std::get<0>(patch.inverse_evaluate(q, u_min - period + d, u_max - period - d, v_min, v_max));
```
Comment (mmorse1217) [inline tests/test_BSplinePatch.cpp:277]: My only reservation with having a tuple returned is the more cluttered interface with std::get<0>(...), but i don't have a better idea. Changing newton-raphson() to a while loop to always converge changes functionality and isn't always desired, so maybe this is best.

File: include/nanospline/PatchBase.h [hunk 2]
```diff
@@ -273,6 +331,10 @@ class PatchBase
             Eigen::Matrix<Scalar, 2, 2> J;
             J << Su.squaredNorm() + r.dot(Suu), Su.dot(Sv) + r.dot(Suv), Sv.dot(Su) + r.dot(Suv),
                 Sv.squaredNorm() + r.dot(Svv);
+            if (J.determinant() <= 0) {
+                uv = prev_uv;
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (qnzhou) [reply include/nanospline/PatchBase.h:337]: Thanks for pointing it out.  I will add the optimality conditions as the stopping criteria.  The Hessian of the shortest distance function is not always positive definite.  I think the Newton iterations is surely not a on a convergence path if a negative `|J|` is detected, thus stopping early make sense.

File: tests/test_BSplinePatch.cpp (hunk 1 above)
Comment (qnzhou) [reply tests/test_BSplinePatch.cpp:277]: Yup, sorry about the API change.  Another option is to make either uv or the convergence flag an argout parameter.  But argout parameters are also not very nice.

File: include/nanospline/PatchBase.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mmorse1217) [reply include/nanospline/PatchBase.h:343]: I can't find the higher dimensional version of the theorem off hand, but [here's the 1D version ](https://en.wikipedia.org/wiki/Newton%27s_method#Analysis). The relevant part:

> If f is continuously differentiable and its derivative is nonzero at α, then there exists a neighborhood of α such that for all starting values x0 in that neighborhood, the sequence {xn} will converge to α.

I think (but I need to double check) that in more general cases, `|J|>0` doesn't imply a lack of convergence, but that the current guess at the solution is simply not in the region of quadratic convergence. Newton will converge linearly until it reaches this region.
When using sampling-based or subpatch-based approaches to get an initial guess, we're pretty confident that the guess in the quadratic convergence zone when they terminate for manifold geometries. If the user provides an initial guess, though, we can't guarantee that they land in the quadratic convergence region. I think that, in this case, breaking when `|J| < 0` could skip a trajectory along which the solution may eventually have converged.

[the remark this answers is not in the document]
Comment (qnzhou) [reply include/nanospline/PatchBase.h:343]: > I think (but I need to double check) that in more general cases, |J| < 0 doesn't imply a lack of convergence along the entire trajectory, but only that the current guess at the solution is not in the region of quadratic convergence. Newton can converge linearly until the current guess lands in this region.

Interesting, I thought that when `|J| < 0`, convergence, if ever reached, tends to be by accident or luck.

**Pull request merged**